lib/rules/no-sync.js

```diff
@@ -19,22 +19,79 @@ module.exports = {
             recommended: false
         },
 
-        schema: []
+        schema: [
+            {
+                type: "object",
+                properties: {
+                    allowAtRootLevel: {
+                        type: "boolean"
+                    }
+                },
+                additionalProperties: false
+            }
+        ]
     },
 
     create(context) {
+        const allowAtRootLevel = context.options[0] ? context.options[0].allowAtRootLevel : false;
+        const stackOfSyncExpressions = [];
+        let setOfSyncExpressions = new Set();
 
-        return {
+        /**
+         * push Sync Expression sets and track Sync Expressions in a new set
+         *
+         * @returns {void}
+         */
+        function trackSyncExpression() {
+            if (allowAtRootLevel) {
+                stackOfSyncExpressions.push(setOfSyncExpressions);
+                setOfSyncExpressions = new Set();
+            }
+        }
 
-            "MemberExpression[property.name=/.*Sync$/]"(node) {
-                context.report({
-                    node,
-                    message: "Unexpected sync method: '{{propertyName}}'.",
-                    data: {
-                        propertyName: node.property.name
-                    }
-                });
+        /**
+         * Report if the Sync Expression is not in the root level
+         *
+         * @param {ASTNode} node the FunctionDeclaration/Expression node
+         * @returns {void}
+         */
+        function reportIfSyncInFunction(node) {
+            if (allowAtRootLevel) {
+                if (setOfSyncExpressions.size > 0) {
+                    setOfSyncExpressions.forEach(syncExpression => {
+                        context.report({
+                            node,
+                            message: "Unexpected sync method: '{{propertyName}}'.",
+                            data: {
+                                propertyName: syncExpression.property.name
+                            }
+                        });
+                    });
+
+                }
+                setOfSyncExpressions.clear();
+                setOfSyncExpressions = stackOfSyncExpressions.pop();
             }
+        }
+
+        return {
+            "MemberExpression[property.name=/.*Sync$/]"(node) {
```

Comment:
Instead of manually keeping track of function scope, I think it would be simpler to use a [descendant AST selector](http://eslint.org/docs/developer-guide/selectors#what-syntax-can-selectors-have):

```js
module.exports = {
    meta: {
        // ...
    },
    create(context) {
        const selector = context.options[0] && context.options[0].allowAtRootLevel
            ? ":function MemberExpression[property.name=/.*Sync$/]"
            : "MemberExpression[property.name=/.*Sync$/]"

        return {
            [selector](node) {
                context.report({
                    node,
                    message: "Unexpected sync method: '{{propertyName}}'.",
                    data: {
                        propertyName: node.property.name
                    }
                });
            }
        }
    }
};
```

That way you won't have to do any of the function scope tracking yourself.